File: 236.Lowest Common Ancestor of a Binary Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    // // brute force: O(N)time& space, TLE for large data
    // TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
    //     vector<TreeNode*> path_p = findPath(root, p, vector<TreeNode*>());
    //     vector<TreeNode*> path_q = findPath(root, q, vector<TreeNode*>());
    //     int i = 0, j = 0;
    //     while(i < path_p.size() && j < path_q.size()){
    //         if(path_p[i] == path_q[j]){
    //             i++; j++;
    //         }
    //         else return path_p[i - 1];
    //     }
    //     return i == path_p.size()? path_p[i - 1]: path_q[j - 1];
    // }

    // vector<TreeNode*> findPath(TreeNode* root, TreeNode* t, vector<TreeNode*> path){
    //     if(!root) return vector<TreeNode*>();
    //     path.push_back(root);
    //     if(root == t) return path;
    //     vector<TreeNode*> left = findPath(root->left, t, path);
    //     vector<TreeNode*> right = findPath(root->right, t, path);
    //     return left.empty()? right: left;
    // }

    // recursive: O(N)time, O(1)space, 53.10%, ok
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        if(!root || root == p || root == q) return root;
        TreeNode* left = lowestCommonAncestor(root->left, p, q);
        TreeNode* right = lowestCommonAncestor(root->right, p, q);
        if(left && right) return root;
        return !left? right: left;
    }
};

File: 236. Lowest Common Ancestor of a Binary Tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    
#     # dfs find path & hashmap check LCA: O(N) time, O(2^N) space, MLE
#     def lowestCommonAncestor(self, root, p, q):
#         """
#         :type root: TreeNode
#         :type p: TreeNode
#         :type q: TreeNode
#         :rtype: TreeNode
#         """
#         def dfs(root, target, path, ret):
#             if root == target:
#                 ret += path + [root] 
#                 return 
#             if root.left:
#                 dfs(root.left, target, path + [root], ret)
#             if root.right:
#                 dfs(root.right, target, path + [root], ret)
        
#         path_p, path_q = [], []
#         dfs(root, p, [], path_p)
#         dfs(root, q, [], path_q)
#         path_p = {node:1 for node in path_p}
#         for node in path_q[::-1]:
#             if node in path_p:
#                 return node
#         return None

#     # improved, dfs find path & hashmap check LCA: O(N) time, O(2^N) space, MLE
#     def lowestCommonAncestor(self, root, p, q):
#         def dfs(root, target, path, ret):
#             if root == target:
#                 ret += path + [root] 
#                 return 
#             if root.left:
#                 dfs(root.left, target, path + [root], ret)
#             if root.right:
#                 dfs(root.right, target, path + [root], ret)
        
#         def findLCA(root, target, nodes, curr):
#             if root in nodes:
#                 curr = root
#             if root == target:
#                 self.ret = curr
#                 return 
#             for children in filter(None, [root.left, root.right]):
#                 findLCA(children, target, nodes, curr)
        
#         path_p = []
#         dfs(root, p, [], path_p)        
#         path_p = {node:1 for node in path_p}
        
#         self.ret = None
#         findLCA(root, q, path_p, root)
#         return self.ret

    # dfs, if subtree t has both p and q, t is LCA: O(N) time, O(N) space
    def lowestCommonAncestor(self, root, p, q):
        if root in [None, p, q]:
            return root
        left = self.lowestCommonAncestor(root.left, p, q)
        right = self.lowestCommonAncestor(root.right, p, q)
        if left and right:
            return root
        return left if right is None else right
            